odbc/src/api/odbc_uri.rs

```diff
@@ -19,23 +38,79 @@ impl<'a> ODBCUri<'a> {
         if odbc_uri.is_empty() {
             return Err(ODBCError::InvalidUriFormat(NOT_EMPTY_ERROR.to_string()));
         }
-        // TODO SQL-990: Support the actual ODBC spec with regards to special characters in attributes
-        // the algorithm will most likely need to be a state machine over each character in the
-        // odbc_uri string.
-        odbc_uri
-            .split(';')
-            .filter(|attr| !attr.is_empty())
-            .map(|attr| {
-                // now split each attribute pair on '='
-                let mut sp = attr.split('=').collect::<Vec<_>>();
-                if sp.len() != 2 {
-                    return Err(ODBCError::InvalidUriFormat(EQUAL_ERROR.to_string()));
-                }
-                // ODBC attribute keys are case insensitive, so we lowercase the keys
-                Ok((sp.remove(0).to_lowercase(), sp.remove(0)))
-            })
-            .collect::<Result<HashMap<_, _>>>()
-            .map(ODBCUri)
+        let mut input = odbc_uri;
+        let mut ret = ODBCUri(HashMap::new());
+        while let Some((keyword, value, rest)) = ODBCUri::get_next_attribute(input)? {
+            if keyword.is_empty() {
+                return Ok(ret);
+            }
+            ret.0.insert(keyword, value);
+            if rest.is_none() {
+                return Ok(ret);
+            }
+            input = rest.unwrap();
+        }
+        Ok(ret)
+    }
+
+    fn get_next_attribute(odbc_uri: &'a str) -> Result<Option<(String, &'a str, Option<&'a str>)>> {
+        // clean up any extra semi-colons
+        let index = odbc_uri.find(|c| c != ';');
+        // these are just trailing semis on the URI
+        if index.is_none() {
+            return Ok(None);
+        }
+        let odbc_uri = odbc_uri.get(index.unwrap()..).unwrap();
+        // find the first '=' sign, '=' does not appear in any keywords, so this is safe.
+        let (keyword, rest) = odbc_uri.split_at(
+            odbc_uri
+                .find('=')
+                .ok_or(ODBCError::InvalidUriFormat(EQUAL_ERROR.to_string()))?,
+        );
+        // remove the leading '=' sign.
+        let rest = rest.get(1..).unwrap();
+        if !KEYWORDS.is_match(keyword) {
+            return Err(ODBCError::InvalidUriFormat(format!(
+                "'{}' is not a valid URI keyword",
+                keyword
+            )));
+        }
+        let (value, rest) = if rest.chars().nth(0).unwrap() == '{' {
+            let rest = rest.get(1..).ok_or_else(|| {
+                ODBCError::InvalidUriFormat(MISSING_CLOSING_BRACE_ERROR.to_string())
+            })?;
+            ODBCUri::handle_braced_value(rest)?
+        } else {
+            ODBCUri::handle_unbraced_value(rest)?
+        };
+        Ok(Some((keyword.to_lowercase(), value, rest)))
+    }
+
+    fn handle_braced_value(input: &'a str) -> Result<(&'a str, Option<&'a str>)> {
```

Comment:
I'm actually not sure now if we should, perhaps, allow nested {}. What do you guys think? It wouldn't be that hard to support since the braced_value is its own function, but it isn't at all clear from the spec if we should.

@nbagnard, what do you think?

Reply:
I did not see it specified in the spec, I vote for not supporting it unless a customer asks.

Reply:
> A DSN or connection string value enclosed with braces ({}) containing any of the characters []{}(),;?*=!@ is passed intact to the driver. However, when using these characters in a keyword, the Driver Manager returns an error when working with file DSNs but passes the connection string to the driver for regular connection strings. Avoid using embedded braces in a keyword value.

To me, this implies attribute values _can_ include `{}` so we should be supporting them. Maybe I'm misinterpreting this spec, though.

Reply:
I think you are right!